Write docs for ESM Views (#1704)
jpmorganchase/modular: docs/concepts/microfrontends.md

---
parent: Concepts
---

## ESM micro frontends in Modular

Micro frontends are a pattern in which discrete UIs (frontends) are composed of
independent fragments that can be built and deployed separately by different
teams and loaded on-demand at runtime. Modular gives developers the opportunity
to implement micro frontends through [ESM Views](../building-apps/esm-views.md),
which are applications built as ES Modules. ESM Views can be served standalone
or dynamically imported by a host application at runtime and rendered in the
host application's own React tree, without the need of using Iframes. ESM Views

> also allow automatic dependency de-duplication, thanks to how we offload

Reviewer: ```suggestion
also allow automatic dependency de-duplication in the browser, thanks to how we offload
```

---
parent: Concepts
---

## ESM micro frontends in Modular

> Micro frontends are a pattern where discrete UIs (frontends) are composed from

Author: Done

independent fragments that can be built and deployed separately by different
teams and loaded on-demand at runtime. Modular gives developers the opportunity

> of implementing microfrontends through

Author: Grammar, Consistency: Done
Content: I changed the sentence and made "ESM Views" its subject. Hope it's clearer.

> [ESM views](../building-apps/esm-views.md), which are applications built as ES

Author: +1 for consistency. Done.

Modules, that can be served standalone or dynamically imported by an host

> application. This is particularly useful when a View exports a default React

Author: > There's a difference between a "default React component" and "exports a React component as its default export" - use the latter.
> Do developers have to export a React component as their default export?

This doesn't apply anymore - I changed the sentence to be less specific.

> Do developers have to export a React component as their default export? What if they want to export multiple things? I would start with a simple Hello World example and then mention how alternatives can be handled, if they can.

We're kinda still defining this; right now, the user can export whatever they want, but the trampoline expects a React component as default export. This means that if the user doesn't export a React component, the built ES Module will work perfectly _as an ES Module_, but serving the dist directory in standalone mode won't work. We still don't know what users will prefer / ask, so I wouldn't be too prescriptive here.

> application. This is particularly useful when a View exports a default React

Author: (If you remember, we were discussing with @steveukx that in the future, depending on how users actually need to use esm-views, we might have a flag in the manifest configuration to trigger the creation of index + trampoline)

> application. This is particularly useful when a View exports a default React

Reviewer 2: Okay, so given that we are releasing this functionality in a somewhat beta state, we should make it clear that some of this is subject to change.

> application. This is particularly useful when a View exports a default React

Author: > Okay, so given that we are releasing this functionality in a somewhat beta state, we should make it clear that some of this is subject to change.

We have a technical description of how it works [in the `esm views.md` file](https://github.com/jpmorganchase/modular/blob/558990b8638c396a384aa52adffcf201374cf44e/docs/building-apps/esm%20views.md?plain=1#L60) already: that reference file might / will change as we refine our technical approach. In this `microfrontend.md` file I'm trying to give a more bird's eye view of the topic; it doesn't matter *how* we import an esm-view; what matters is that we can import it at runtime without re-downloading its dependencies or create an iframe for it. I removed the "default export" part here as it's not so relevant and added it to `esm views.md`.

> application. This is particularly useful when a View exports a default React

Reviewer 2: We can call out that this is subject to change wherever you think is best. We don't want to surprise users who are depending on functionality that changes.

> component: an host application can import it at runtime and render it in its own

Author: Done

React tree, without the need of using Iframes and with automatic dependency

> de-duplication, thanks to how we offload dependencies to an external ESM CDN.

Author: Modular. Done.

## How we build micro frontends

A package of `type: "esm-view"` is built as an ES Module by the `modular build`
command. The view's local dependencies are bundled in the dist result, while the
view's external dependencies are rewritten to import from an ESM CDN by default,
like (for example) [Skypack](https://www.skypack.dev/) or
[esm.sh](https://esm.sh/). This allows the view's external dependencies to be

> loaded on-demand at runtime and to be automatically de-duplicated (i.e. the same

Author: Not really. Cache decides if browser needs to download a module, but there is also a mechanism to ensure that modules are not re-evaluated in the current page context which is independent from cache hit.

dependency will be not re-fetched and re-evaluated, but simply re-used if any

> view tries to use it the second time) and it's especially useful in a React

Author: Done.

micro frontend scenario, where multiple ESM views share the same copy of React,

> without incurring in errors associated with

Author: Done.

[multiple copies of React in the same page](https://reactjs.org/warnings/invalid-hook-call-warning.html).

Since dynamic import of ESM Modules leverages a
[widely supported standard](https://caniuse.com/es6-module-dynamic-import),
where the browser does the heavy lifting of fetching, de-duplicating and making

> dependencies available, ESM Modules are modular's choice building blocks to

Author: done.

implement a micro frontend architecture.

## How to load micro frontends

ESM views, when built, generate a single javascript entrypoint and a single CSS

> entrypoint that can be imported ar runtime by any application (or other view)

Author: Done. ESM Views.

using dynamic import (or any viable technique in case of styles). ESM views will

> generate a package manifest (`package.json` file) which contains information

Author: Done.

regarding the location of the built files (whose names are uniquely hashed to
facilitate caching) and lists of bundled and rewritten dependencies along with
their dependencies, in order to decouple importing of ESM views from the actual
build result structure. For more information, visit the
[ESM Views reference page](../building-apps/esm-views.md)

## Standalone support

> As ESM views typically export a React component by default, a synthetic

Author: It will and, in that case, it will break. We discussed this already and said that in the future we might want a configuration flag for the synthetic index / trampoline.

`index.html` capable of loading the view as a standalone web page is provided in
the `dist/{{view-name}}` directory. This means that ESM views can additionally

> be served (for example with the `modular serve` command, or any HTTP server) as

Author: I liked the comma with an hat, but just removed the comma.

```json
"modular": {
    "type": "esm-view"
}
```

This will instruct the `modular build` command to output an
[ES Module](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules).

The ESM View's local dependencies will be bundled in the dist result, while the
ESM View's external dependencies will be rewritten to import from an ESM CDN by
default, like (for example) [Skypack](https://www.skypack.dev/) or
[esm.sh](https://esm.sh/). This allows the ESM View's external dependencies to
be loaded on-demand at runtime and to be automatically de-duplicated (i.e. the
same dependency will be not re-fetched and re-evaluated, but simply re-used if
any view tries to use it the second time). This is especially useful in a React
micro frontend scenario, where multiple ESM views share the same copy of React,
without incurring errors associated with
[multiple copies of React in the same page](https://reactjs.org/warnings/invalid-hook-call-warning.html).

Since dynamically importing ES Modules leverages a
[widely supported standard](https://caniuse.com/es6-module-dynamic-import),
where the browser does the heavy lifting of fetching, de-duplicating and making

> dependencies available, ES Modules are Modular's choice building blocks to

Reviewer: ```suggestion
dependencies available, ES Modules are Modular's preferred building blocks to
```

implement a micro frontend architecture.

## How to load micro frontends

> ESM Views, when built, generate a single JavaScript entrypoint and a single CSS

Reviewer: Simplified clause...

```suggestion
Building an ESM View generates a single JavaScript entrypoint and a single CSS
```
